### PR TITLE
(feature): Add :company-kind to eglot-completion-at-point

### LINKED ISSUE #652: Feature: Support candidate-kind backend call for completion-at-point function
Forked from [company#1070](https://github.com/company-mode/company-mode/pull/1070)

Company recently added support for configuring a prefix for completion candidates in the company popup (see [company#756](https://github.com/company-mode/company-mode/issues/756)). The APIs also been updated to allow you to ask what type (in this case LSP kind) a candidate is. It'd be nice if eglot conformed to this new API change :smile:.

### PR DESCRIPTION
**Pull request by mohkale**
Closes #652

### CODE REVIEW

**Comment by mohkale** [comment]
@joaotavora 

I ended up using `intern` instead of `make-symbol` cause I'm not sure enough on the difference between them and intern has always worked well enough for me. Would you like me to change it to `make-symbol` instead?

**Comment by joaotavora** [comment]
> intern has always worked well enough for me. 

eh, yes, probably you need intern.  here's the difference
```
(eq (intern "foo") (intern "foo")) ;; => T

(eq (make-symbol "foo") (make-symbol "foo")) ;; => nil
```

So `intern` always finds the "interned" symbol object, and returns that object.  It only makes a new symbol object with the given string as name if it doesn't find it "interned". `make-symbol` always does a new symbol object.

**Comment by mohkale** [comment]
oh, thanks for the clarification :smile:. I'll keep it as intern for now then.

**Comment by dgutov** [comment]
Anything I can help with to speed this along?

**Comment by joaotavora** [comment]
Thanks. You can review it, clean up long lines, make the commit look follow Emacs's commit message guidelines, confirm the copyright assignment situation, etc, etc.  I've given you commit rights.

**Comment by dgutov** [comment]
Not exactly the help I was thinking of, but let's try to organize this. The assignment is already in place.

@joaotavora Do you have some contributing instructions here? E.g. ones that mention ChangeLog style log messages.

**Comment by joaotavora** [comment]
> @joaotavora Do you have some contributing instructions here? E.g. ones that mention ChangeLog style log messages.

Just the same as Emacs's.  Whatever goes for Emacs goes here. Because this is still github start the subject line like

`Close #656: add frobnicating kittens`

But even that is not important.  And yes, @mohkale already has a FSF assignment in place.

**Comment by dgutov** [comment]
In any case, @mohkale, please reformat the code so each line is <80 characters long, and use the Change Log style for the commit message. Example: https://www.gnu.org/prep/standards/html_node/Style-of-Change-Logs.html#Style-of-Change-Logs (you can also look at the existing commit history in this project).

**Comment by dgutov** [comment]
> Just the same as Emacs's. Whatever goes for Emacs goes here.

Even if you just say this, that still needs to be written somewhere, right?

**Comment by joaotavora** [comment]
> Even if you just say this, that still needs to be written somewhere, right?

Right, I suppose so.  You can create that file minimally if you want.

**Comment by mohkale** [comment]
> In any case, @mohkale, please reformat the code so each line is <80 characters long, and use the Change Log style for the commit message. Example: https://www.gnu.org/prep/standards/html_node/Style-of-Change-Logs.html#Style-of-Change-Logs (you can also look at the existing commit history in this project).

@dgutov

I've made the changes you asked for (noticed that the changelog style doesn't match the one described above but I ended up using the one suggested by joatavora), also one of the comments is an org-mode link (I insert them sporadically across everything) and it pushes the line length upto 128. Would you like me to remove it as well?

**Comment by dgutov** [comment]
> noticed that the changelog style doesn't match the one described above but I ended up using the one suggested by joatavora

His suggestion was for the summary line (which is free-form), then you add an empty line, and then you follow with a changelog-style entry like

```
* eglot.el (aasd): New function.
(asadfasdf): Use it.
```

> also one of the comments is an org-mode link (I insert them sporadically across everything)

Do they display differently in your Emacs? Like, linkified?

My preference would be to keep the comment but remove the formatting. Like this:

```
       ;; Associate each lsp-item with a lsp-kind symbol, should be one of these:
       ;; https://github.com/company-mode/company-mode/blob/de80d17cd2e7e51826c72ef85d94d38a0fca0a76/company.el#L1365.
```

**Comment by joaotavora** [comment]
> My preference would be to keep the comment but remove the formatting. Like this:

Yes, this is common simple Emacs style. 

Also  maybe just pointing to the company file and saying the symbol is enough, and not as brittle should company migrate away from github.  Just a suggestion.

File: eglot.el [hunk 1]
```diff
@@ -2236,6 +2236,15 @@ is not active."
                (concat " "
                        (propertize annotation
                                    'face 'font-lock-function-name-face))))))
+       :company-kind
+       ;; Associate each lsp-item with a lsp-kind symbol, should be one of
+       ;; [[https://github.com/company-mode/company-mode/blob/de80d17cd2e7e51826c72ef85d94d38a0fca0a76/company.el#L1365][these]].
+       (lambda (proxy)
```
Comment (joaotavora) [inline eglot.el:2242]: Is the function not guaranteed to be called with non-nil?  Maybe @dgutov knows this.

**Comment by mohkale** [comment]
> His suggestion was for the summary line (which is free-form), then you add an empty line, and then you follow with a changelog-style entry like

oh, my bad. I've updated it now.

> Do they display differently in your Emacs? Like, linkified?

They do when you've got `org-link-minor-mode` enabled thankfully (it's part of the org-contrib package).

File: eglot.el (hunk 1 above)
Comment (dgutov) [reply eglot.el:2242]: Yeah, I think you can remove the `when`.

**Comment by joaotavora** [comment]
> They do when you've got org-link-minor-mode enabled thankfully (it's part of the org-contrib package). If you've got ace-link or link-hint you can also jump to and open links using avy.

But not everybody has that, and that includes the maintainer of this package, so please just say that it matches some variable in company.  Or don't even say that, as that's an external implementation detail anyway which we don't want to marry ourselves to.  We're marrying ourselves to `:company-kind` which probably is documented independently.

**Comment by mohkale** [comment]
@joaotavora 

Roger, done :-).

File: eglot.el (hunk 1 above)
Comment (dgutov) [reply eglot.el:2241]: What you *could* be checking for, is whether `proxy` is an empty string. I don't know if Eglot can return those, but some weird backends do.

Comment (dgutov) [reply eglot.el:2241]: Ah, never mind. `get-text-property` doesn't blow up on those (anymore?)

**Comment by dgutov** [comment]
@mohkale Look into removing the `when` too, please (see the subthread).

File: eglot.el (hunk 1 above)
Comment (mohkale) [reply eglot.el:2241]: I've removed the when now :smile:.

**Comment by dgutov** [comment]
Aand in it goes.

Thanks for the patience!

**Comment by joaotavora** [comment]
I prefer rebase, but the occasional merge is fine too.  Thanks @dgutov for the guidance and @mohkale for the code.

**Comment by dgutov** [comment]
I didn't want to make it seem, from the log, like the commit needed further tweaks.

**Comment by joaotavora** [comment]
> I didn't want to make it seem, from the log, like the commit needed further tweaks.

Don't follow, but it's fine like this. And tweaking someone else's commit or commit message is also fine as long as attribution is kept (just add yourself as `Co-authored-by:`).  Again, I think these are the same rules as in Emacs, or at least I behave similarly in Emacs and here.

**Pull request merged**